packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+      recoloringPerRound[0] = 0;
+      vertsPerRound[0] = 0;
+      incorrectGhostsPerRound[0]=0;
+      typename Kokkos::View<int*, device_type>::HostMirror ghost_colors_host;
+      typename Kokkos::View<lno_t*, device_type>::HostMirror boundary_verts_host;
+      int serial_threshold = this->pl->template get<int>("serial_threshold",0);
+      //see if recoloring is necessary.
+      gno_t totalConflicts = 0;
+      gno_t localConflicts = recoloringSize_host(0);
+      Teuchos::reduceAll<int,gno_t>(*comm, Teuchos::REDUCE_SUM, 1, &localConflicts, &totalConflicts);
+      bool done = !totalConflicts;
+      if(comm->getSize()==1) done = true;
+      
+      //recolor until no conflicts are left
+      while(!done){
+	if(recoloringSize_host(0) < serial_threshold) break;
+        if(distributedRounds < numStatisticRecordingRounds) {
+          vertsPerRound[distributedRounds] = verts_to_recolor_size_host(0);
+        }
+        
+        comm->barrier();
+        double recolor_temp = timer();
+        //recolor using KokkosKernels' coloring function 
+        if(verts_to_recolor_size_host(0) > 0){
+	  this->colorInterior(femv_colors.size(), dist_adjs_dev, dist_offsets_dev,femv,verts_to_recolor_view,verts_to_recolor_size_host(0),use_vbbit);
+	}
+
+	if(distributedRounds < numStatisticRecordingRounds){
+          recoloringPerRound[distributedRounds] = timer() - recolor_temp;
+          recoloring_time += recoloringPerRound[distributedRounds];
+          total_time += recoloringPerRound[distributedRounds];
+          comp_time += recoloringPerRound[distributedRounds];
+          compPerRound[distributedRounds] = recoloringPerRound[distributedRounds];
+          totalPerRound[distributedRounds] = recoloringPerRound[distributedRounds];
+	}
+
+	Kokkos::parallel_for(n_ghosts, KOKKOS_LAMBDA(const int& i){
+          femv_colors(i+n_local) = ghost_colors(i);
+        });
+        Kokkos::fence();
+        
+	comm->barrier();
+	//send views are up-to-date, they were copied after conflict detection.
+        //communicate the new colors
+        double curr_comm_time = doOwnedToGhosts(mapOwnedPlusGhosts,n_local,verts_to_send_host,verts_to_send_size_host,femv,procs_to_send,sent,recv);
+	comm_time += curr_comm_time;
+
+        if(distributedRounds < numStatisticRecordingRounds){
+          commPerRound[distributedRounds] = curr_comm_time;
+	  recvPerRound[distributedRounds] = recv;
+          sentPerRound[distributedRounds] = sent;
+	  if(verbose) {
+            std::cout<<comm->getRank()<<": total sent in round "<<distributedRounds<<" = "<<sent<<"\n";
+            std::cout<<comm->getRank()<<": total recv in round "<<distributedRounds<<" = "<<recv<<"\n";
+	  }
+          totalPerRound[distributedRounds] += commPerRound[distributedRounds];
+	}
+
+        Kokkos::parallel_for(n_ghosts, KOKKOS_LAMBDA(const int& i){
+          ghost_colors(i) = femv_colors(i+n_local);
+        });
+        Kokkos::fence();
+        comm->barrier();
```

Comment:
Is this barrier needed only for timing?  If so, should it be closer to the timer?